packages/aquedux-client/src/createClient.js

```diff
@@ -0,0 +1,122 @@
+  // Ideally, these actions should not be sent back to client. Duplicates definition in middleware
+  function isWiredActionType(actionType: string): boolean {
+    const internalActionTypes = [
+      actionTypes.AQUEDUX_CLIENT_CHANNEL_JOIN,
+      actionTypes.AQUEDUX_CLIENT_CHANNEL_LEAVE,
+      actionTypes.AQUEDUX_CHANNEL_SNAPSHOT
+    ]
+
+    return internalActionTypes.indexOf(actionType) !== -1 || hydratedActionTypes.indexOf(actionType) !== -1
+  }
+
+  function onSend(action) {
+    // If the socket is not open, we stack the actions.
+    if (!socket || socket.readyState !== 1) {
+      actionStack.push(action)
+      return
+    }
+
+    // Append our jwt from the local storage to the action before sending it.
+    const token = localStorage.getItem('aquedux-jwt')
+    if (token !== null) {
+      action = {
+        ...action,
+        token
+      }
+    }
+
+    socket.send(JSON.stringify(action))
+  }
+
+  function onMessage(socket, message) {
+    const json = (() => {
+      try {
+        return JSON.parse(message.data)
+      } catch (e) {
+        return null
+      }
+    })()
+
+    if (!json || !json.type) {
+      console.error(`The client received an invalid message: ${JSON.stringify({ message: json }, null, 2)}`)
+      return
+    }
+
+    // If the message has a jwt, override our own.
+    if (json.token) {
+      localStorage.setItem('aquedux-jwt', json.token)
+    }
+
+    const action = (() => {
+      if (isWiredActionType(json.type)) {
+        /**
+         * We need to change type to bypass potential user middlewares
+         * Real action.type will be restored in aquedux middleware before hitting reducers
+         */
+        return {
+          ...json,
+          originalActionType: json.type,
+          type: actionTypes.AQUEDUX_CLIENT_MESSAGE_RECEIVED
+        }
+      } else {
+        return json
+      }
+    })()
+
+    raise(events.EVENT_ACTION_RECEIVE, action)
+  }
+
+  function onOpen(isRestart = false) {
+    // Try to resubscribe to previously subscribed channels before loosing connection.
+    if (isRestart) {
+      raise(events.EVENT_CHANNEL_RESUBSCRIBE)
+    } else {
+      // Send stacked actions that occured when the socket was not ready yet.
+      actionStack.forEach(onSend)
+      actionStack = []
+    }
+  }
+
+  function onStart(isRestart = false) {
+    unregister(events.EVENT_CLIENT_START, onStart)
+    register(events.EVENT_CLIENT_STOP, onStop)
+    register(events.EVENT_CLIENT_RESTART, onRestart)
+    register(events.EVENT_ACTION_SEND, onSend)
+
+    socket = new SockJS(endpoint)
+    socket.onopen = () => onOpen(isRestart)
+    socket.onclose = () => onStop(socket)
+    socket.onmessage = e => onMessage(socket, e)
+  }
+
+  function onStop(socket) {
+    socket && socket.close()
+
+    unregister(events.EVENT_ACTION_SEND, onSend)
+    unregister(events.EVENT_CLIENT_STOP, onStop)
+    unregister(events.EVENT_CLIENT_RESTART, onRestart)
+    register(events.EVENT_CLIENT_START, onStart)
```

Comment:
Do we really need this register/unregister flow?
Can we register only one time before calling `onStart()` on client creation?

Reply:
A couple of things here:

First my part:
* I realised that the onSend, onMessage and onStop callbacks do not need `socket` as argument and are actually creating errors when triggered by a stop action (the event will trigger `onStop(undefined)`). I must remove the socket argument to let the closure take care of the binding.

About the register/unregister flow: For onStart and onStop it make the `start` action valid only if the client is stopped (and the `stop` action valid if the client is started). Think of it as a `if-statement free` way of ensuring that behaviour.

For the `onSend` and `onRestart` you are right.

There is only a specific situation that bothers me tho: what if the redux store is destroyed and recreated later ? In that situation we fail to clear the handlers and some callbacks with invalid closure to a destroyed socket object will be called 😔 I don't know any standard `dispose` mechanism in JS that would help. Any idea ?